### PR TITLE
Add Oberon7 arm and gripper

### PR DESCRIPTION
**Pull request by iche033**
Replaces #68 

Depends on https://github.com/ignitionrobotics/ign-gazebo/pull/1405 and https://github.com/ignitionrobotics/ign-gazebo/pull/1406

Ported and integrated the oberon7 arm from DAVE / UUV simulator. The arm and gripper model.sdf files are generated using urdf files from:
https://github.com/Field-Robotics-Lab/uuv_manipulators/tree/master/oberon7/oberon7_description
https://github.com/Field-Robotics-Lab/dave/tree/master/urdf/manipulators/dave_oberon7_description

The spawn launch file is updated to take two additional arguments: `arm` and `gripper` to specify the arm and gripper model to use. Currently we only have one arm but we will need be able to swap out the oberon7 gripper with the suction gripper from https://github.com/osrf/mbzirc/pull/63. 

The arm must be spawned with the usv. Currently it only has one mounting position on the usv for the arm. More maybe added in the future.

usv_oberon7_arm_gripper

Usage:

```sh
# launch simple demo world
ros2 launch ros_ign_gazebo ign_gazebo.launch.py ign_args:="-v 4 -r simple_demo.sdf"

# spawn usv with oberon7 arm and oberon7 gripper.
# if gripper is not specified, it'll just use the default one which is the oberon7 gripper
ros2 launch mbzirc_ign spawn.launch.py name:=usv world:=simple_demo model:=usv type:=usv x:=15 y:=0 z:=0.3 R:=0 P:=0 Y:=0 arm:=mbzirc_oberon7_arm gripper:=mbzirc_oberon7_gripper
```

Try publishing topics to control the arm and gripper joints:

```sh
# arm
ros2 topic pub --once /usv/arm/joint/azimuth/cmd_pos std_msgs/msg/Float64 'data: 0.5'
ros2 topic pub --once /usv/arm/joint/shoulder/cmd_pos std_msgs/msg/Float64 'data: 0.5'
ros2 topic pub --once /usv/arm/joint/elbow/cmd_pos std_msgs/msg/Float64 'data: 0.5'
ros2 topic pub --once /usv/arm/joint/roll/cmd_pos std_msgs/msg/Float64 'data: 0.5'
ros2 topic pub --once /usv/arm/joint/pitch/cmd_pos std_msgs/msg/Float64 'data: 0.5'
ros2 topic pub --once /usv/arm/joint/wrist/cmd_pos std_msgs/msg/Float64 'data: 0.5'
# gripper
ros2 topic pub --once /usv/arm/gripper/joint/finger_left/cmd_pos std_msgs/msg/Float64 'data: 0.5'
ros2 topic pub --once /usv/arm/gripper/joint/finger_right/cmd_pos std_msgs/msg/Float64 'data: 0.5'
```

You can also echo the joint states

```sh
ros2 topic echo /usv/arm/joint_states

ros2 topic echo /usv/arm/gripper/joint_states
```

Here's the tf tree:

frames

### CODE REVIEW

**Review by caguero — COMMENTED** [review]
I'm following your instructions but I don't see the arm mounted in the USV when I run: `ros2 launch mbzirc_ign spawn.launch.py name:=usv world:=simple_demo model:=usv type:=usv x:=15 y:=0 z:=0.3 R:=0 P:=0 Y:=0 arm:=mbzirc_oberon7_arm gripper:=mbzirc_oberon7_gripper`

File: mbzirc_ign/src/mbzirc_ign/bridges.py [hunk 1]
```diff
@@ -102,3 +102,41 @@ def thrust_joint_pos(model_name, side):
...
+        ign_type='ignition.msgs.Model',
+        ros_type='sensor_msgs/msg/JointState',
+        direction=BridgeDirection.IGN_TO_ROS)
+
+
+def gripper_joint_states(world_name, model_name):
+    gripper_prefix = f'/world/{world_name}/model/{model_name}/model/arm/model/gripper'
+    return Bridge(
+        ign_topic=f'{gripper_prefix}/joint_state',
+        ros_topic='/joint_states',
```
Comment (caguero) [inline mbzirc_ign/src/mbzirc_ign/bridges.py:121]: Should this topic be the same as the one used for the arm joint states?

**Comment by caguero** [comment]
> ```shell
> ```shell
> ros2 launch mbzirc_ign spawn.launch.py name:=usv world:=simple_demo model:=usv type:=usv x:=15 y:=0 z:=0.3 R:=0 P:=0 Y:=0 arm:=mbzirc_oberon7_arm gripper:=mbzirc_oberon7_gripper
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```

Forget this comment, I can see the arm now after cleaning and rebuilding.

**Review by caguero — COMMENTED** [review]
The fingers of the gripper start completely closed. I can open the fingers but I can't close them entirely anymore.

**Comment by iche033** [comment]
> The fingers of the gripper start completely closed. I can open the fingers but I can't close them entirely anymore.

hmm I'm trying to reproduce this. Here's what I'm doing but I see that the fingers are able to close. What are the joint cmds you're using?

```
# open gripper
ros2 topic pub --once /usv/arm/gripper/joint/finger_right/cmd_pos std_msgs/msg/Float64 'data: 1'
ros2 topic pub --once /usv/arm/gripper/joint/finger_left/cmd_pos std_msgs/msg/Float64 'data: 1'

# close gripper
ros2 topic pub --once /usv/arm/gripper/joint/finger_right/cmd_pos std_msgs/msg/Float64 'data: 0'
ros2 topic pub --once /usv/arm/gripper/joint/finger_left/cmd_pos std_msgs/msg/Float64 'data: 0'
```

oberon7_finger_opening_closing

**Comment by caguero** [comment]
> ros2 topic pub --once /usv/arm/gripper/joint/finger_right/cmd_pos std_msgs/msg/Float64 'data: 1'

Works for me as well. It was doing it in the wrong way.

**Review by caguero — COMMENTED** [review]
Ready to approve, just a minor question about the joint states from the gripper.

File: mbzirc_ign/src/mbzirc_ign/bridges.py (hunk 1 above)
Comment (iche033) [reply mbzirc_ign/src/mbzirc_ign/bridges.py]: I've seen both cases before:

* `/joint_states`
* `/<namespace>/joint_states`

I think other packages like moveit subscribes to `/joint_states` by default but can be configured to subscribe to namespaced topics.

I changed to the form `/<namespace>/joint_states` for the arm and gripper in [00ed7c1](https://github.com/osrf/mbzirc/pull/72/commits/00ed7c1068bc0ab086710c6bddc70313dc8ae4b6). It's now:

`/usv/arm/joint_states`
`/usv/arm/gripper/joint_states`

The user can remap them back to `/joint_states` if they wish.

**Pull request merged**